### PR TITLE
Make (Multi)CheckpointWriter consume StreamingMap

### PR DESCRIPTION
**Pull request by vjeko**
In order for checkpointing to work with large (disk-backed) tables,
(Multi)CheckpointWriter needs to consume the streaming interface.

## Overview

Description:

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

File: test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java [hunk 1]
```diff
@@ -532,13 +533,13 @@ public void checkpointAndTrimAtDifferentPoint() throws Exception {
             final int sizeMap1 = 4;
             final int sizeMap2 = 6;
 
-            Map<String, String> map1 = defaultRT.getObjectsView().build()
+            StreamingMap<String, String> map1 = defaultRT.getObjectsView().build()
```
Comment (hisundar) [inline test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java]: these will involve about 168 changes in upstream since Map<> is used in those places.

Comment (vjeko) [reply test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java]: There is nothing forbidding you from using it as a Map. The only place where this change will be needed is wherever MultiCheckpointWriter is being consumed.

Comment (vjeko) [reply test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java]: In other words, the change in the contract happens here:

`public class MultiCheckpointWriter<T extends StreamingMap> `

Comment (vjeko) [reply test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java]: I checked the corfu-compactor logic, we consume `MultiCheckpointWriter` using `ICorfuMap`:

`MultiCheckpointWriter<ICorfuMap> mcWriter = new MultiCheckpointWriter<>();`

And `ICorfuMap` is defined as:

```
public interface ICorfuMap<K, V>
    extends StreamingMap<K, V> {
```

Thus, from what I can tell no changes should be necessary on the client side.

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 2]
```diff
@@ -40,7 +42,7 @@
  *  TODO: Generalize to all SMR objects.
  */
 @Slf4j
-public class CheckpointWriter<T extends Map> {
+public class CheckpointWriter<T extends StreamingMap> {
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: StreamingMap is too low level to expose on a public API. I think `ICorfuSMR` is more appropriate, `ISMRMap`.

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java [hunk 3]
```diff
@@ -163,18 +165,18 @@ public Token appendCheckpoint(Token snapshotTimestamp, Long streamTail) {
             // A checkpoint writer will do two accesses one to obtain the object
             // vlo version and to get a shallow copy of the entry set
             log.info("appendCheckpoint: Started checkpoint for {} at snapshot {}", streamId, snapshotTimestamp);
-            Set<Map.Entry> entries = this.map.entrySet();
+            Stream<Map.Entry> entries = this.map.entryStream();
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: `ICorfuSMR` can expose a `getStateMethod`, or `ISMRMap` can expose `getEntryStream`

File: runtime/src/main/java/org/corfudb/runtime/collections/ICorfuMap.java [hunk 4]
```diff
@@ -5,7 +5,7 @@
 import java.util.function.Predicate;
 
 public interface ICorfuMap<K, V>
-    extends Map<K, V> {
+    extends StreamingMap<K, V> {
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/collections/ICorfuMap.java]: I think these should be reverse. `ICorfuMap` is the "more general" type.

File: runtime/src/main/java/org/corfudb/runtime/collections/SMRMap.java [hunk 5]
```diff
@@ -31,7 +32,7 @@
 @SuppressWarnings("checkstyle:abbreviation")
 public class SMRMap<K, V>
         extends HashMap<K, V>
-        implements ISMRMap<K,V>, ICorfuSMR<SMRMap<K, V>> {
+        implements StreamingMap<K, V>, ISMRMap<K,V>, ICorfuSMR<SMRMap<K, V>> {
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/collections/SMRMap.java]: The type relationships is already complex.

I think it would be cleaner to have ISMRMap extend ICorfuSMR. How can an object be SMR if it doesn't implement ICorfuSMR, all SMRObject types should extend/implement that type.

File: test/src/test/java/org/corfudb/integration/AbstractIT.java [hunk 6]
```diff
@@ -305,8 +306,8 @@ public static CorfuRuntime createRuntimeWithCache() {
         return rt;
     }
 
-    public static Map<String, Integer> createMap(CorfuRuntime rt, String streamName) {
-        Map<String, Integer> map = rt.getObjectsView()
+    public static StreamingMap<String, Integer> createMap(CorfuRuntime rt, String streamName) {
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/integration/AbstractIT.java]: Why is this needed?

File: test/src/test/java/org/corfudb/integration/LogSizeQuotaIT.java [hunk 7]
```diff
@@ -118,7 +119,7 @@ public void testLogUnitQuota() throws Exception {
         privilegedRt.connect();
 
 
-        Map<String, String> map2 = privilegedRt.getObjectsView()
+        StreamingMap<String, String> map2 = privilegedRt.getObjectsView()
```
Comment (Maithem) [inline test/src/test/java/org/corfudb/integration/LogSizeQuotaIT.java]: Why is this needed?

File: test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java (hunk 1 above)
Comment (Maithem) [reply test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java]: MultiCheckpointWriter shouldn't even use generics since we never need them.

I think its more appropriate to introduce new tests to test `StreamingMap` and not change existing ones.

**Comment by Maithem** [comment]
I think its better to introduce new tests for the new functionality and not overload what's currently there.

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java (hunk 2 above)
Comment (vjeko) [reply runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: Consumed via `ISMRMap` now.

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java (hunk 3 above)
Comment (vjeko) [reply runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: Consumed via `ISMRMap` now.

File: runtime/src/main/java/org/corfudb/runtime/collections/SMRMap.java (hunk 5 above)
Comment (vjeko) [reply runtime/src/main/java/org/corfudb/runtime/collections/SMRMap.java]: I was trying to avoid the diamond problem that arises due to `AutoClosable` interface. One solution would be to force all concrete implementation of `ICorfuSMR` to implement `close()`, but due those tricky annotation we have (`@PassThrough @DontInstrument`), I am hesitant to go down that path.

File: test/src/test/java/org/corfudb/integration/AbstractIT.java (hunk 6 above)
Comment (vjeko) [reply test/src/test/java/org/corfudb/integration/AbstractIT.java]: `Checkpointer` needs to consume an interface that exposes the streaming API, so `Map` will have to be replaced with something else. Most likely `ISMRMap`.

File: test/src/test/java/org/corfudb/integration/LogSizeQuotaIT.java (hunk 7 above)
Comment (vjeko) [reply test/src/test/java/org/corfudb/integration/LogSizeQuotaIT.java]: Take a look at the previous comment: https://github.com/CorfuDB/CorfuDB/pull/2237#discussion_r351501861

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java (hunk 2 above)
Comment (vjeko) [reply runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: Turns out we cannot use `ISMRMap`, since this interface is only consumed only `SMRMap` and not `CorfuTable`. Making `CorfuTable` inherit from `ISMRMap` makes little sense as it contains bunch of logic only specific to `SMRMap`.

If we were to remove `SMRMap`, this change ought to be possible. For now I ensured that `StreamingMap` does not expose any Corfu specific internal -- it just contains `Stream<Map.Entry<K, V>> entryStream()`.

File: runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java (hunk 3 above)
Comment (vjeko) [reply runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java]: You look at the previous comment regarding `CorfuSMR`.

As for `ICorfuSMR` exposing `getStateMethod` -- I entertained that idea and I concluded that the complexity is not worth the effort. In order to achieve this, we would have to add and extra generic parameter to `ICorfuSMR` `E`, which in case of `CorfuTable` would be `Stream<Map.Entry<K, V>>`, but in other cases it (`CorfuCompoundObject`, `CorfuSharedCounter`) would not be a sequence. Also, after new GC comes in, all of this code is going to be irrelevant.

File: runtime/src/main/java/org/corfudb/runtime/collections/ICorfuMap.java (hunk 4 above)
Comment (vjeko) [reply runtime/src/main/java/org/corfudb/runtime/collections/ICorfuMap.java]: `StreamingMap` is not a simplified interface.

File: test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java (hunk 1 above)
Comment (vjeko) [reply test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java]: You technically don't need the generic parameter. In this case it serves more as a type alias.

The tests have not been changed, just the upper bound type associated with the constructed objects.
